service/kernel/data_management/train_test_split.cpp

```diff
@@ -0,0 +1,203 @@
+        const size_t end   = iBlock == nTrainBlocks - 1 ? start + BLOCK_SIZE + nTrainSurplus : start + BLOCK_SIZE;
+
+        for (size_t i = start; i < end; ++i)
+            for (size_t j = 0; j < nColumns; ++j)
+                trainDataPtr[i * nColumns + j] = origDataPtr[trainIdx[i] * nColumns + j];
+    });
+
+    runBlocks(nTestRows * nColumns > THREADING_BORDER, nTestBlocks, [&](size_t iBlock) {
+        const size_t start = iBlock * BLOCK_SIZE;
+        const size_t end   = iBlock == nTestBlocks - 1 ? start + BLOCK_SIZE + nTestSurplus : start + BLOCK_SIZE;
+
+        for (size_t i = start; i < end; ++i)
+            for (size_t j = 0; j < nColumns; ++j)
+                testDataPtr[i * nColumns + j] = origDataPtr[testIdx[i] * nColumns + j];
+    });
+    return s;
+}
+
+
+template <typename IdxType>
+services::Status trainTestSplitImpl(NumericTable & inputTable, NumericTable & trainTable, NumericTable & testTable,
+                                    NumericTable & trainIdxTable, NumericTable & testIdxTable, NumericTable & columnTypesTable)
+{
+    services::Status s;
+    const NTLayout layout  = inputTable.getDataLayout();
+    const size_t nColumns = trainTable.getNumberOfColumns();
+    const size_t nTrainRows = trainTable.getNumberOfRows();
+    const size_t nTestRows = testTable.getNumberOfRows();
+    DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(size_t, nTrainRows + nTestRows, nColumns);
+
+    BlockDescriptor<int> columnTypesBlock;
+    columnTypesTable.getBlockOfRows(0, nColumns, readOnly, columnTypesBlock);
+    const int * columnTypesPtr = columnTypesBlock.getBlockPtr();
+    DAAL_CHECK_MALLOC(columnTypesPtr);
+
+    BlockDescriptor<IdxType> trainIdxBlock, testIdxBlock;
+    trainIdxTable.getBlockOfColumnValues(0, 0, nTrainRows, readOnly, trainIdxBlock);
+    testIdxTable.getBlockOfColumnValues(0, 0, nTestRows, readOnly, testIdxBlock);
+    const IdxType * trainIdx = trainIdxBlock.getBlockPtr();
+    const IdxType * testIdx = testIdxBlock.getBlockPtr();
+    DAAL_CHECK_MALLOC(trainIdx);
+    DAAL_CHECK_MALLOC(testIdx);
+
+    if (layout == NTLayout::soa)
+    {
+        runBlocks(nColumns > 1 && nColumns * (nTrainRows + nTestRows) > THREADING_BORDER, nColumns, [&](size_t iCol) {
+            switch (columnTypesPtr[iCol])
+            {
+                case 0:
+                    splitColumn<int, IdxType>(inputTable, trainTable, testTable, trainIdx, testIdx, nTrainRows, nTestRows, iCol);
+                    break;
+                case 1:
+                    splitColumn<float, IdxType>(inputTable, trainTable, testTable, trainIdx, testIdx, nTrainRows, nTestRows, iCol);
+                    break;
+                case 2:
+                    splitColumn<double, IdxType>(inputTable, trainTable, testTable, trainIdx, testIdx, nTrainRows, nTestRows, iCol);
+                    break;
+            }
+        });
+    }
+    else
+    {
+        switch (columnTypesPtr[0])
+        {
+            case 0:
```

Comment:
What do 0,1,2 mean exactly? I suppose it's values in some enum

Reply:
Changes to IndexNumType enum